File: app/src/main/java/com/nirbhor/app/ui/theme/Theme.kt
package com.nirbhor.app.ui.theme

import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.remember
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight

/**
 * Reads the Nirbhor design tokens. Use [NirbhorTheme.colors], [NirbhorTheme.type], plus the
 * [Dimens] / [NirbhorShapes] objects, in place of raw hexes and sizes everywhere.
 */
object NirbhorTheme {
    val colors: NirbhorColors
        @Composable get() = LocalNirbhorColors.current
    val type: NirbhorType
        @Composable get() = LocalNirbhorType.current
}

/**
 * App theme. Bangla is primary, so [isBangla] defaults true. [biggerText] applies the "সহজে পড়ার
 * জন্য / bigger text" accessibility setting (up to ~1.3× per the responsive spec).
 *
 * The palette is a single warm-light scheme; the design does not define a dark variant, so we keep
 * the calm health palette in both system themes rather than inverting it.
 */
@Composable
fun NirbhorTheme(
    isBangla: Boolean = true,
    biggerText: Boolean = false,
    content: @Composable () -> Unit,
) {
    @Suppress("UNUSED_VARIABLE") // reserved: a future dark palette could branch on this.
    val dark = isSystemInDarkTheme()
    val colors = remember { NirbhorColors() }
    val scale = if (biggerText) 1.15f else 1f
    val type = remember(isBangla, scale) { buildNirbhorType(isBangla, scale) }

    val defaultFamily: FontFamily = if (isBangla) AnekBangla else Archivo

    // Material3 scheme, mapped from our tokens so stock M3 components stay on-palette.
    // error → warm (never red).
    val scheme = lightColorScheme(
        primary = colors.calm,
        onPrimary = colors.paper,
        primaryContainer = colors.calmSoft,
        onPrimaryContainer = colors.calmD,
        secondary = colors.calm,
        onSecondary = colors.paper,
        secondaryContainer = colors.sage,
        onSecondaryContainer = colors.ink,
        background = colors.paper,
        onBackground = colors.ink,
        surface = colors.card,
        onSurface = colors.ink,
        surfaceVariant = colors.sage,
        onSurfaceVariant = colors.ink2,
        outline = colors.line,
        outlineVariant = colors.line,
        error = colors.warm,
        onError = colors.paper,
        errorContainer = colors.warmSoft,
        onErrorContainer = colors.warmD,
    )

    val m3Typography = remember(type) {
        val base = Typography()
        val bodyStyle = type.body.copy(fontFamily = defaultFamily)
        base.copy(
            bodyLarge = bodyStyle,
            bodyMedium = bodyStyle,
            bodySmall = type.meta.copy(fontFamily = defaultFamily),
            titleLarge = type.header.copy(fontFamily = defaultFamily, fontWeight = FontWeight(700)),
            labelLarge = type.buttonLabel.copy(fontFamily = defaultFamily),
        )
    }

    CompositionLocalProvider(
        LocalNirbhorColors provides colors,
        LocalNirbhorType provides type,
    ) {
        MaterialTheme(
            colorScheme = scheme,
            typography = m3Typography,
            content = content,
        )
    }
}

File: lib/theme/theme.dart
import 'package:flutter/material.dart';

import '../i18n/numerals.dart';
import 'colors.dart';
import 'typography.dart';

export 'colors.dart';
export 'dimens.dart';
export 'typography.dart';

/// Reads the Nirbhor design tokens and locale primitives. Compose held these in composition-locals;
/// here one [InheritedWidget] carries the same four values down the tree.
///
/// Bangla is primary. Rather than a central strings file (a merge bottleneck when many screens are
/// built at once), screens pair the two languages at the call site with [NbContext.tr] — which
/// mirrors how the design pairs bn/en copy. Numerals/time formatting read the same values.
class NirbhorTheme extends InheritedWidget {
  const NirbhorTheme({
    super.key,
    required this.type,
    required this.isBangla,
    required this.is24Hour,
    required super.child,
  });

  final NirbhorType type;
  final bool isBangla;
  final bool is24Hour;

  static NirbhorTheme of(BuildContext context) {
    final t = context.dependOnInheritedWidgetOfExactType<NirbhorTheme>();
    assert(t != null, 'NirbhorTheme is missing above this widget');
    return t!;
  }

  @override
  bool updateShouldNotify(NirbhorTheme old) =>
      type != old.type || isBangla != old.isBangla || is24Hour != old.is24Hour;
}

/// The token accessors every widget uses in place of raw hexes and sizes.
extension NbContext on BuildContext {
  NirbhorColors get colors => nbColors;
  NirbhorType get type => NirbhorTheme.of(this).type;
  bool get isBangla => NirbhorTheme.of(this).isBangla;
  bool get is24Hour => NirbhorTheme.of(this).is24Hour;

  /// Picks Bangla or English copy for the active locale.
  String tr(String bn, String en) => isBangla ? bn : en;

  /// Localised integer (Bengali numerals in the Bangla locale).
  String num(int value) => Numerals.number(value, isBangla);

  /// Localises the digits in an already-formatted string (e.g. "৫৪টি").
  String numStr(String text) => Numerals.digits(text, isBangla);

  String percent(int value) => Numerals.percent(value, isBangla);

  String qty(double value) => Numerals.quantity(value, isBangla);

  /// Localised clock time for the active locale + time-format.
  String clock(int hour, int minute) => Numerals.time(hour, minute, isBangla, is24Hour);
}

/// Material theme mapped from our tokens so stock Material components stay on-palette.
/// `error` maps to warm (never red).
///
/// The palette is a single warm-light scheme; the design does not define a dark variant, so we keep
/// the calm health palette in both system themes rather than inverting it.
ThemeData buildMaterialTheme(NirbhorType type, bool isBangla) {
  const c = nbColors;
  final family = isBangla ? kAnekBangla : kArchivo;
  return ThemeData(
    useMaterial3: true,
    fontFamily: family,
    scaffoldBackgroundColor: c.paper,
    splashFactory: InkRipple.splashFactory,
    colorScheme: ColorScheme(
      brightness: Brightness.light,
      primary: c.calm,
      onPrimary: c.paper,
      primaryContainer: c.calmSoft,
      onPrimaryContainer: c.calmD,
      secondary: c.calm,
      onSecondary: c.paper,
      secondaryContainer: c.sage,
      onSecondaryContainer: c.ink,
      surface: c.card,
      onSurface: c.ink,
      surfaceContainerHighest: c.sage,
      onSurfaceVariant: c.ink2,
      outline: c.line,
      outlineVariant: c.line,
      error: c.warm,
      onError: c.paper,
      errorContainer: c.warmSoft,
      onErrorContainer: c.warmD,
    ),
    textTheme: TextTheme(
      bodyLarge: type.body,
      bodyMedium: type.body,
      bodySmall: type.meta,
      titleLarge: type.header,
      labelLarge: type.buttonLabel,
    ),
  );
}
